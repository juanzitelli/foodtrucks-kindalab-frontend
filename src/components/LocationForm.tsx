import _ from 'lodash';
import React, { ChangeEvent, ChangeEventHandler, useState } from 'react'
import PlacesAutocomplete, { getLatLng, geocodeByAddress } from 'react-places-autocomplete'
import { useDispatch } from 'react-redux';
import styled, { CSSProperties } from 'styled-components';
import { startFetchingFoodTrucks } from './../redux/actions/foodtrucks.actions';

const LocationForm = () => {
	const [address, setAddress] = useState("")
	const [limit, setLimit] = useState(10)
	const dispatch = useDispatch();

	const handleAddressSelect = async (value: string) => {
		const results = await geocodeByAddress(value)
		const { lat, lng } = await getLatLng(results[0])
		setAddress(value);
		dispatch(startFetchingFoodTrucks(lat, lng, limit))
	}

	const handleLimitChange: ChangeEventHandler = (event: ChangeEvent<HTMLInputElement>) => {
		setLimit(_.toNumber(event.target.value))
	}

	const LocationInputStyles: CSSProperties = {
		padding: "0.8rem 3rem",
		width: "30vw",
		fontSize: "1.1rem",
	}
	const LimitInputStyles: CSSProperties = {
		padding: "0.8rem 3rem",
		width: "10vw",
		fontSize: "1.1rem",
		marginLeft: "1rem"
	}

	const SuggestionStyles = {
		paddingTop: "5px",
		cursor: "pointer"
	}
	const SuggestionContainer = styled.div`
		display: flex;
		flex-direction: column;
		justify-items: flex-start;
		align-items: flex-start;
		margin-top: 1rem;
	`

	const divStyles = {
		marginLeft: "3rem",
		marginTop: "3rem",
	}

	return (
		<div style={{ height: "100%" }}>
			<PlacesAutocomplete value={address} onChange={setAddress} onSelect={handleAddressSelect}>
				{({ getInputProps, suggestions, getSuggestionItemProps, loading }) => (
					<div style={divStyles}>
						<input style={LocationInputStyles}{...getInputProps({ placeholder: "Type an address" })} />
						<input style={LimitInputStyles} onChange={handleLimitChange} value={limit} type="number" name="limit" id="limit" />
						{loading ? <span>...Loading</span> : null}
						<SuggestionContainer>
							{
								suggestions.map(suggestion => {
									return (<span style={SuggestionStyles} {...getSuggestionItemProps(suggestion)}>{suggestion.description}</span>)
								})
							}
						</SuggestionContainer>
					</div>
				)
				}
			</PlacesAutocomplete>
		</div>
	)
}

export default LocationForm
