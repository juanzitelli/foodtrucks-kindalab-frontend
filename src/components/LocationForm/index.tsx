import _ from 'lodash'
import React, { useEffect } from 'react'
import { geocodeByAddress, getLatLng } from 'react-places-autocomplete';
import { useDispatch } from 'react-redux';
import useForm from '../../hooks/useForm';
import { useTypedSelector } from '../../hooks/useTypedSelector';
import { startFetchingFoodTrucks } from '../../redux/actions/foodtrucks.actions';
import { predictionsCleared, predictionSelected, startFetchingPlaces } from '../../redux/actions/places.actions';
import { Prediction as PredictionType } from '../../types/Prediction';
import { StyledForm } from './styles/StyledForm'
import { FormGroup } from './styles/FormGroup'
import Prediction from './components/Prediction';
import { StyledInput } from './styles/StyledInput';
import { StyledButton } from './styles/StyledButton';
import { PredictionContainer } from './styles/PredictionsContainer';

const LocationForm = () => {

	const placesPredictionsLimit = 5

	const dispatch = useDispatch()

	type LocationFormData = {
		location: string;
		limit: number;
	}

	const initialState: LocationFormData = {
		location: "",
		limit: 3
	}

	const [formValues, handleInputChange, resetForm, changeValue] = useForm<LocationFormData>(initialState)

	const { location, limit } = formValues;

	const { predictions, selectedPrediction } = useTypedSelector(state => state.places)
	const { foodTrucksLoaded } = useTypedSelector(state => state.ui)

	useEffect(() => {
		dispatch(startFetchingPlaces(location, placesPredictionsLimit))
	}, [location, limit, dispatch])

	const formSubmitHandler = async (event: React.SyntheticEvent) => {
		event.preventDefault();
		const results = await geocodeByAddress(location)
		const { lat, lng } = await getLatLng(results[0])	
		
		dispatch(startFetchingFoodTrucks(lat, lng, limit))
		
	}

	const predictionClickHandler = (prediction: PredictionType) => {
		changeValue('location', prediction.description)
		dispatch(predictionSelected(prediction))
		dispatch(predictionsCleared())
	}

	const clearLocationHandler = () => {
		resetForm();
		dispatch(predictionsCleared())
	}

	return <>
		<StyledForm onSubmit={formSubmitHandler}>
			<FormGroup>
				<label htmlFor="location">Location</label>
				<div>
					<StyledInput placeholder="1250 5th Avenue, San Francisco" autoComplete={`off`} type="text" name="location" id="location" value={location} onChange={handleInputChange} />
					<StyledButton onClick={clearLocationHandler}>Clear</StyledButton>
				</div>
			</FormGroup>

			<PredictionContainer hasPredictions={!_.isUndefined(predictions)}>
				{
					!_.isUndefined(predictions) && (predictions as PredictionType[]).map(prediction => {
						return prediction ? <Prediction prediction={prediction} onClickHandler={() => predictionClickHandler(prediction)} /> : null
					})
				}
			</PredictionContainer>

			<FormGroup>
				<label htmlFor="limit">Results quantity</label>
				<StyledInput type="number" name="limit" id="limit" min={1} value={limit} onChange={handleInputChange} />
			</FormGroup>
			{
				foodTrucksLoaded ? <div>
					{
						selectedPrediction ? <StyledButton type="submit">Search</StyledButton> : <p>Find and select a location to search for food trucks!</p>
					}
				</div> : <p>...Loading</p>
			}
		</StyledForm>
	</>
}

export default LocationForm
