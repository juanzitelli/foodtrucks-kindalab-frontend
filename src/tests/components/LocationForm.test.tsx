import React from 'react';
import { mount, shallow } from 'enzyme';
import { Provider } from 'react-redux';
import thunk from "redux-thunk";
import configureStore from "redux-mock-store";
import LocationForm from '../../components/LocationForm/index';

const middlewares = [thunk];
const createMockStore = configureStore(middlewares);
const initState = {
	foodTrucks: []
};
const mockStore = createMockStore(initState)
mockStore.dispatch = jest.fn();

describe('Tests on LocationForm component', () => {
	const wrapper = shallow(
		<Provider store={mockStore}>
			<LocationForm />
		</Provider>)
	test('Should match snapshot', async () => {
		expect(wrapper).toMatchSnapshot();
	})
})
