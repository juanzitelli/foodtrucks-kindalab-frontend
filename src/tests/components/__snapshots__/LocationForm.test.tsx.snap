// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests on LocationForm component Should match snapshot 1`] = `
<form
  className="sc-bdfBwQ cnWvbD"
  onSubmit={[Function]}
>
  <div
    className="sc-gsTCUz jXKszp"
  >
    <label
      htmlFor="location"
    >
      Location
    </label>
    <div>
      <input
        autoComplete="off"
        className="sc-hKgILt ePrkBg"
        id="location"
        name="location"
        onChange={[Function]}
        placeholder="1250 5th Avenue, San Francisco"
        type="text"
        value=""
      />
      <button
        className="sc-eCssSg gtCJVU"
        onClick={[Function]}
      >
        Clear
      </button>
    </div>
  </div>
  <div
    className="sc-jSgupP krqAwW"
  />
  <div
    className="sc-gsTCUz jXKszp"
  >
    <label
      htmlFor="limit"
    >
      Results quantity
    </label>
    <input
      className="sc-hKgILt ePrkBg"
      id="limit"
      min={1}
      name="limit"
      onChange={[Function]}
      type="number"
      value={3}
    />
  </div>
  <p>
    ...Loading
  </p>
</form>
`;
